src/typecheck/mod.rs

```diff
@@ -76,10 +80,494 @@ use error::*;
+pub type Environment = GenericEnvironment<Ident, UnifType>;
 
 /// Mapping from wildcard ID to inferred type
 pub type Wildcards = Vec<Types>;
+/// Unification variable or type constants unique identifier.
+pub type VarId = usize;
+
+/// A unifiable record row.
+pub type GenericUnifRecordRow<E> = RecordRowF<Box<GenericUnifType<E>>>;
+
+/// Unifiable record rows. Same shape as [`crate::types::RecordRows`], but where each type is
+/// unifiable, and each tail may be a unification variable (or a constant).
+#[derive(Clone, PartialEq, Debug)]
+pub enum GenericUnifRecordRows<E: TermEnvironment + Clone> {
+    Concrete(RecordRowsF<Box<GenericUnifType<E>>, Box<GenericUnifRecordRows<E>>>),
+    Constant(VarId),
+    UnifVar(VarId),
+}
+
+/// Unifiable enum rows. Same shape as [`crate::types::EnumRows`] but where each tail may be a
+/// unification variable (or a constant).
+///
+/// Enum rows don't store any type (they are just a sequence of identifiers), so there is no
+/// `GenericUnifEnumRows` taking an additioinal `E` parameter.
+#[derive(Clone, PartialEq, Debug)]
+pub enum UnifEnumRows {
+    Concrete(EnumRowsF<Box<UnifEnumRows>>),
+    Constant(VarId),
+    UnifVar(VarId),
+}
+
+/// The types on which the unification algorithm operates, which may be either a concrete type, a
+/// type constant or a unification variable.
+///
+/// Contracts store an additional term environment for contract equality checking, which is
+/// represented by `E`. The typechecker always uses the same type for `E`. However, the evaluation
+/// phase may also resort to checking contract equality, using a different environment
+/// representation, hence the parametrization.
+#[derive(Clone, PartialEq, Debug)]
+pub enum GenericUnifType<E: TermEnvironment> {
+    /// A concrete type (like `Num` or `Str -> Str`).
+    Concrete(TypeF<Box<GenericUnifType<E>>, GenericUnifRecordRows<E>, UnifEnumRows>),
+    /// A contract, seen as an opaque type. In order to compute type equality between contracts or
+    /// between a contract and a type, we need to carry an additional environment. This is why we
+    /// don't reuse the variant from [`crate::types::TypeF`].
+    Contract(RichTerm, E),
+    /// A rigid type constant which cannot be unified with anything but itself.
+    Constant(VarId),
+    /// A unification variable.
+    UnifVar(VarId),
+}
+
+impl<E: TermEnvironment + Clone> std::convert::TryInto<RecordRows> for GenericUnifRecordRows<E> {
+    type Error = ();
+
+    fn try_into(self) -> Result<RecordRows, ()> {
+        match self {
+            GenericUnifRecordRows::Concrete(rrows) => {
+                let converted: RecordRowsF<Box<Types>, Box<RecordRows>> = rrows.try_map(
+                    |uty| Ok(Box::new(GenericUnifType::try_into(*uty)?)),
+                    |urrows| {
+                        let rrows: RecordRows = (*urrows).try_into()?;
+                        Ok(Box::new(rrows))
+                    },
+                )?;
```

Comment:
Yes, I thought about that, and then did differently, don't remember why. I'll give it a shot.

Reply:
Now I remember: because the type parameters of the functor versions are generic, you need to know how to traverse the recursive unfolding like `Ty` and `RRows`. To do that, you can either take a function that tell how to traverse them, but then you're just re-implementing `map` under a different name. Or you need traits to tell you that `Ty`, `RRows` and `ERows` are stuff that can be traversed. Because they are respectively a ternary, binary and unary functor, you need three different traits. All of this starts to be a lot of boilerplate, while the definition of subst (while, indeed, redundant with traverse), is not that bad, I think.

Reply:
Yes, that sounds like keeping the current version is best. Let's not turn this PR into an exercise in importing Traversable into rust.